services/insights-handler/internal/handler/main.go

```diff
@@ -0,0 +1,749 @@
+	}
+
+	environment, err := lagoonclient.GetEnvironmentFromName(context.TODO(), apiClient, resource.Environment, project.Id)
+	if err != nil {
+		return lagoonclient.Project{}, lagoonclient.Environment{}, err
+	}
+	return project, environment, nil
+}
+
+func (h *Messaging) sendToLagoonS3(incoming *InsightsMessage, insights InsightsData, resource ResourceDestination) (err error) {
+	// strip http/s protocol from origin
+	u, _ := url.Parse(h.S3Config.S3Origin)
+	if u.Scheme == "http" || u.Scheme == "https" {
+		h.S3Config.S3Origin = u.Host
+	}
+
+	// Push to s3 bucket
+	minioClient, err := minio.New(h.S3Config.S3Origin, &minio.Options{
+		Creds:  credentials.NewStaticV4(h.S3Config.AccessKeyId, h.S3Config.SecretAccessKey, ""),
+		Secure: h.S3Config.UseSSL,
+	})
+	if err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	err = minioClient.MakeBucket(ctx, h.S3Config.Bucket, minio.MakeBucketOptions{Region: h.S3Config.Region})
+	if err != nil {
+		exists, errBucketExists := minioClient.BucketExists(ctx, h.S3Config.Bucket)
+		if errBucketExists != nil && !exists {
+			return err
+		}
+	} else {
+		log.Printf("Successfully created %s\n", h.S3Config.Bucket)
+	}
+
+	if incoming.Payload != nil {
+		b, err := json.Marshal(incoming)
+		if err != nil {
+			return err
+		}
+
+		objectName := strings.ToLower(fmt.Sprintf("%s-%s-%s-%s.json", insights.InputType, resource.Project, resource.Environment, resource.Service))
+		reader := bytes.NewReader(b)
+
+		info, putObjErr := minioClient.PutObject(ctx, h.S3Config.Bucket, objectName, reader, reader.Size(), minio.PutObjectOptions{
+			ContentEncoding: "application/json",
+		})
+		if putObjErr != nil {
+			return putObjErr
+		}
+
+		log.Printf("Successfully uploaded %s of size %d\n", objectName, info.Size)
+	}
+
+	if incoming.BinaryPayload != nil {
+		for _, p := range incoming.BinaryPayload {
+			result, err := decodeString(p)
+			if err != nil {
+				fmt.Errorf(err.Error())
+			}
+			resultJson, _ := json.MarshalIndent(result, "", " ")
+
+			var objectName string
+			for _, f := range insights.S3OutputFormat {
+				objectName = strings.ToLower(fmt.Sprintf("%s-%s-%s-%s-%s.%s", insights.LagoonType, strings.Replace(f, ".", "-", -1), resource.Project, resource.Environment, resource.Service, f))
+
+				tempFilePath := fmt.Sprintf("/tmp/%s", objectName)
+				contentType := fmt.Sprintf("application/%s", strings.Replace(f, ".", "-", -1))
+
+				if f == "json" {
+					err = ioutil.WriteFile(tempFilePath, resultJson, 0644)
+					if err != nil {
+						fmt.Errorf(err.Error())
+					}
+				}
+				if f == "json.gz" {
+					var buf bytes.Buffer
+					gz := gzip.NewWriter(&buf)
+					gz.Write(resultJson)
+					gz.Close()
+					err = ioutil.WriteFile(tempFilePath, buf.Bytes(), 0644)
+					if err != nil {
+						fmt.Errorf(err.Error())
+					}
+				}
+
+				s3FilePath := strings.ToLower(fmt.Sprintf("%s/%s/%s/%s/%s", insights.InputType, resource.Project, resource.Environment, resource.Service, objectName))
```

Comment:
maybe call the top level folder `insights` and store all insights in the same folder?
```suggestion
				s3FilePath := strings.ToLower(fmt.Sprintf("insights/%s/%s/%s", resource.Project, resource.Environment, objectName))
```